turbo/jsonrpc/eth_call.go

```diff
@@ -332,100 +336,167 @@ func (api *APIImpl) EstimateGas(ctx context.Context, argsOrNil *ethapi2.CallArgs
+	}
+	proof.AccountProof = *(*[]hexutility.Bytes)(unsafe.Pointer(&accountProof))
+
+	// get account data from the trie
+	acc, _ := proofTrie.GetAccount(crypto.Keccak256(address.Bytes()))
+	if acc == nil {
+		for i, k := range storageKeys {
+			proof.StorageProof[i] = accounts.StorProofResult{
+				Key:   k,
+				Value: new(hexutil.Big),
+				Proof: nil,
 			}
-			tx = batch
-		} else {
-			loader = trie.NewFlatDBTrieLoader("eth_getProof", rl, nil, nil, false)
 		}
+		return proof, nil
+	}
 
-		reader, err := rpchelper.CreateStateReader(ctx, tx, blockNrOrHash, 0, api.filters, api.stateCache, "")
-		if err != nil {
-			return nil, err
+	proof.Balance = (*hexutil.Big)(acc.Balance.ToBig())
+	proof.Nonce = hexutil.Uint64(acc.Nonce)
+	proof.CodeHash = acc.CodeHash
+	proof.StorageHash = acc.Root
+
+	// if storage is not empty touch keys and build trie
+	if proof.StorageHash.Cmp(libcommon.BytesToHash(commitment.EmptyRootHash)) != 0 && len(storageKeys) != 0 {
+		// touch storage keys
+		for _, storageKey := range storageKeys {
+			sdCtx.TouchKey(kv.StorageDomain, string(common.FromHex(address.Hex()[2:]+storageKey.String()[2:])), nil)
 		}
-		a, err := reader.ReadAccountData(address)
+
+		// generate the trie for proofs, this works by loading the merkle paths to the touched keys
+		proofTrie, _, err = sdCtx.Witness(ctx, header.Root[:], "eth_getProof")
 		if err != nil {
 			return nil, err
 		}
-		if a == nil {
-			a = &accounts.Account{}
+	}
+
+	tx, err := api.db.BeginTemporalRo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+	reader, err := rpchelper.CreateStateReader(ctx, tx, api._blockReader, blockNrOrHash, 0, api.filters, api.stateCache, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// get storage key proofs
+	for i, keyHash := range storageKeys {
+		proof.StorageProof[i].Key = keyHash
+
+		// if we have simple non contract account just set values directly without requesting any key proof
+		if proof.StorageHash.Cmp(libcommon.BytesToHash(commitment.EmptyRootHash)) == 0 {
+			proof.StorageProof[i].Proof = nil
+			proof.StorageProof[i].Value = new(hexutil.Big)
+			continue
 		}
-		pr, err := trie.NewProofRetainer(address, a, storageKeys, rl)
+
+		// prepare key path (keccak(address) | keccak(key))
+		var fullKey []byte
+		fullKey = append(fullKey, crypto.Keccak256(address.Bytes())...)
+		fullKey = append(fullKey, crypto.Keccak256(keyHash.Bytes())...)
+
+		// get proof for the given key
+		storageProof, err := proofTrie.Prove(fullKey, len(proof.AccountProof), true)
 		if err != nil {
-			return nil, err
+			return nil, errors.New("cannot verify store proof")
 		}
 
-		loader.SetProofRetainer(pr)
-		root, err := loader.CalcTrieRoot(tx, nil)
+		res, err := reader.ReadAccountStorage(address, acc.Incarnation, &keyHash)
 		if err != nil {
-			return nil, err
+			res = []byte{}
+			fmt.Printf("couldn't read account storage for the address %s\n", address.String())
+		}
+		n := new(big.Int)
+		n.SetString(hex.EncodeToString(res), 16)
```

Comment:
wait, if res is a slice of bytes, why convert it to string then parse string representation, can say `n.SetBytes(res)`